### PR TITLE
Add withdraw endpoint to API v1.1

### PR DESCRIPTION
**Pull request by JR-G**
## Context

In v1.1 of the API we're introducing the ability for providers to withdraw (or decline) applications.

This PR adds the `/withdraw` endpoint and the required functionality.

## Changes proposed in this pull request

- Adding the `/withdraw` endpoint
- Also moved the `ProviderInterface::DeclineOrWithdrawApplication` service up a folder level
- Updated the v1.1. API specs to remove the `withdrawn_at` field

## Link to Trello card

https://trello.com/c/SVkHWOg2
https://trello.com/c/21is4HDo

## Things to check

- [x] If the code removes any existing feature flags, a data migration has also been added to delete the entry from the database
- [x] This code does not rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] Required environment variables have been updated [added to the Azure KeyVault](/docs/environment-variables.md#deploy-pipeline)

### CODE REVIEW

File: app/controllers/vendor_api/withdraw_or_decline_controller.rb [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    def withdraw_or_decline
+      application = ProviderInterface::DeclineOrWithdrawApplication.new(
+        actor: audit_user,
+        application_choice: application_choice,
+      )
+
+      if application.save!
+        render_application
+      else
+        invalid_transition
```
Comment (JR-G) [inline app/controllers/vendor_api/withdraw_or_decline_controller.rb:16]: temp

File: app/controllers/vendor_api/withdraw_or_decline_controller.rb [hunk 2]
```diff
@@ -0,0 +1,49 @@
+module VendorAPI
+  class WithdrawOrDeclineController < VendorAPIController
```
Comment (despo) [inline app/controllers/vendor_api/withdraw_or_decline_controller.rb:2]:
```suggestion
  class WithdrawOrDeclineOfferController < VendorAPIController
```

File: app/controllers/vendor_api/withdraw_or_decline_controller.rb [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    before_action :validate_metadata!
+
+    include ApplicationDataConcerns
+
+    def withdraw_or_decline
```
Comment (despo) [inline app/controllers/vendor_api/withdraw_or_decline_controller.rb:7]:
```suggestion
    def create
```

File: spec/requests/vendor_api/v1.1/post_withdraw_or_decline_application_spec.rb [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    application_choice = create_application_choice_for_currently_authenticated_provider(
+      status: 'awaiting_provider_decision',
+    )
+
+    post_api_request "/api/v1.1/applications/#{application_choice.id}/withdraw"
+
+    expect(response).to have_http_status(:ok)
+    application = parsed_response['data']['attributes']
+    expect(application['status']).to eq 'withdrawn'
+    expect(application['withdrawn_or_declined_for_candidate_by_provider']).to eq true
```
Comment (cwrw) [inline spec/requests/vendor_api/v1.1/post_withdraw_or_decline_application_spec.rb:21]: could you also add an expectation for the timestamp being set at withdrawn/declined?

File: app/presenters/concerns/withdraw_or_decline_application_api_data.rb [hunk 5]
```diff
@@ -0,0 +1,9 @@
+module WithdrawOrDeclineApplicationAPIData
+  def schema
+    super.deep_merge!({
+      attributes: {
+        withdrawn_or_declined_for_candidate_by_provider: application_choice.withdrawn_or_declined_for_candidate_by_provider,
```
Comment (cwrw) [inline app/presenters/concerns/withdraw_or_decline_application_api_data.rb:5]: Could we surface to providers the field names `withdrawn_or_declined_for_candidate` as set in the documentation?

Comment (cwrw) [reply app/presenters/concerns/withdraw_or_decline_application_api_data.rb:5]: So I realised we have a withdrawal object with a timestamp, could we update the docs to remove `withdrawn_at` which we added to the draft spec and point to the withdrawal object, and update the specs please?

**Review by cwrw — COMMENTED** [review]
looks great, some minor comments inline

File: app/presenters/concerns/withdraw_or_decline_application_api_data.rb (hunk 5 above)
Comment (JR-G) [reply app/presenters/concerns/withdraw_or_decline_application_api_data.rb]: updated spec

**Review by cwrw — APPROVED** [review]
👍

**Pull request merged**